python/cudf/cudf/core/indexed_frame.py

```diff
@@ -756,6 +757,47 @@ def add_suffix(self, suffix):
                 Use `Series.add_suffix` or `DataFrame.add_suffix`"
         )
 
+    @annotate("APPLY", color="purple", domain="cudf_python")
+    def _apply(self, func, kernel_getter, *args, **kwargs):
+        """Apply `func` across the rows of the frame."""
+        if kwargs:
+            raise ValueError("UDFs using **kwargs are not yet supported.")
+
+        try:
+            kernel, retty = _compile_or_get(
+                self, func, args, kernel_getter=kernel_getter
+            )
+        except Exception as e:
+            raise ValueError(
+                "user defined function compilation or execution failed."
```

Comment:
Is the function executed as part of the compilation pipeline?

Reply:
The python code is never executed. It's definitely "execution" in airquotes here. Let me know if this seems like it would cause confusion.

Reply:
I'm not sure what you mean by execution in this comment. Could something other than compilation have failed?

Reply:
I believe there's situations where a kernel successfully compiles but then there's some kind of runtime error during launch. I'd be worried about this even more as we start leveraging external functions written in c++.

Reply:
> I'm not sure what you mean by execution in this comment. Could something other than compilation have failed?

+1, I think if runtime error is raised during kernel launch, should we instead wrap the `kernel.forall` line in try except block and move phrases of `execution failed` over there?

Reply:
Oh I see what you're saying. Yes, absolutely.

Reply:
I think Michael hit the nail on the head. If the failure is at the execution level, that warning belongs where the kernel is actually executed, not here. That's what I was trying to get at.

Reply:
This should be updated.